File: android/app/src/main/java/com/iptv/M3UParser.kt
package com.iptv

import java.io.BufferedReader
import java.io.StringReader

class M3UParser {
    
    fun parseM3U(content: String): List<Channel> {
        val channels = mutableListOf<Channel>()
        val reader = BufferedReader(StringReader(content))
        
        var line: String?
        var currentChannel: Channel? = null
        
        while (reader.readLine().also { line = it } != null) {
            line?.let { currentLine ->
                when {
                    currentLine.startsWith("#EXTINF:") -> {
                        currentChannel = parseExtInf(currentLine)
                    }
                    currentLine.startsWith("http") && currentChannel != null -> {
                        currentChannel?.let { channel ->
                            channels.add(channel.copy(url = currentLine.trim()))
                        }
                        currentChannel = null
                    }
                }
            }
        }
        
        return channels
    }
    
    private fun parseExtInf(line: String): Channel {
        var name = ""
        var category = ""
        var logo = ""
        
        // Extrair nome do canal
        val nameMatch = Regex(",(.+)$").find(line)
        name = nameMatch?.groupValues?.get(1)?.trim() ?: "Canal Desconhecido"
        
        // Extrair categoria
        val categoryMatch = Regex("group-title=\"([^\"]+)\"").find(line)
        category = categoryMatch?.groupValues?.get(1) ?: "Geral"
        
        // Extrair logo
        val logoMatch = Regex("tvg-logo=\"([^\"]+)\"").find(line)
        logo = logoMatch?.groupValues?.get(1) ?: ""
        
        return Channel(
            name = name,
            url = "",
            category = category,
            logo = logo
        )
    }
    
    fun generateM3U(channels: List<Channel>): String {
        val builder = StringBuilder()
        builder.appendLine("#EXTM3U")
        
        channels.forEach { channel ->
            builder.appendLine("#EXTINF:-1 group-title=\"${channel.category}\" tvg-logo=\"${channel.logo}\",${channel.name}")
            builder.appendLine(channel.url)
        }
        
        return builder.toString()
    }
}

File: ios/IPTVPlayer/Models/M3UParser.swift
import Foundation

class M3UParser {
    
    func parseM3U(content: String) -> [Channel] {
        var channels: [Channel] = []
        let lines = content.components(separatedBy: .newlines)
        
        var currentChannel: Channel?
        
        for line in lines {
            let trimmedLine = line.trimmingCharacters(in: .whitespacesAndNewlines)
            
            if trimmedLine.hasPrefix("#EXTINF:") {
                currentChannel = parseExtInf(line: trimmedLine)
            } else if trimmedLine.hasPrefix("http") && currentChannel != nil {
                if var channel = currentChannel {
                    let finalChannel = Channel(
                        name: channel.name,
                        url: trimmedLine,
                        category: channel.category,
                        logo: channel.logo
                    )
                    channels.append(finalChannel)
                }
                currentChannel = nil
            }
        }
        
        return channels
    }
    
    private func parseExtInf(line: String) -> Channel {
        var name = "Canal Desconhecido"
        var category = "Geral"
        var logo = ""
        
        // Extrair nome do canal
        if let nameRange = line.range(of: ",") {
            name = String(line[nameRange.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
        }
        
        // Extrair categoria
        if let categoryMatch = line.range(of: "group-title=\"") {
            let startIndex = categoryMatch.upperBound
            if let endIndex = line[startIndex...].range(of: "\"")?.lowerBound {
                category = String(line[startIndex..<endIndex])
            }
        }
        
        // Extrair logo
        if let logoMatch = line.range(of: "tvg-logo=\"") {
            let startIndex = logoMatch.upperBound
            if let endIndex = line[startIndex...].range(of: "\"")?.lowerBound {
                logo = String(line[startIndex..<endIndex])
            }
        }
        
        return Channel(name: name, url: "", category: category, logo: logo)
    }
    
    func generateM3U(channels: [Channel]) -> String {
        var content = "#EXTM3U\n"
        
        for channel in channels {
            content += "#EXTINF:-1 group-title=\"\(channel.category)\" tvg-logo=\"\(channel.logo)\",\(channel.name)\n"
            content += "\(channel.url)\n"
        }
        
        return content
    }
}